sailfishusermanager.cpp

```diff
@@ -0,0 +1,429 @@
+        return false;
+    }
+
+    foreach (const QString& file, source.entryList(QDir::Files | QDir::NoDotAndDotDot | QDir::Hidden)) {
+        QString destFile = QString("%1/%2").arg(destination).arg(file);
+        if (!QFile::copy(QString("%1/%2").arg(source.path()).arg(file), destFile)) {
+            qWarning() << "Failed to copy file";
+            return false;
+        }
+        if (chown(destFile.toLocal8Bit(), pw->pw_uid, pw->pw_gid)) {
+            qWarning() << "Failed to change file ownership";
+            return false;
+        }
+    }
+
+    return true;
+}
+
+void SailfishUserManager::addUser(const QString& user, const QString& name, const QDBusMessage& message)
+{
+    m_timer->start();
+
+    if (QFile::exists(QString(USER_HOME).arg(user))) {
+        qWarning() << "Home directory already exists";
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Home directory already exists"));
+        return;
+    }
+
+    if (!addGroup(user)) {
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Creating user group failed"));
+        return;
+    }
+
+    struct lu_error *error = nullptr;
+    struct lu_context *context = lu_start(NULL, lu_user, NULL, NULL, lu_prompt_console, NULL, &error);
+    if (!context) {
+        qWarning() << "Error creating context:" << lu_strerror(error);
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
+        return;
+    }
+
+    struct lu_ent *ent_user = lu_ent_new();
+    if (lu_user_default(context, user.toLocal8Bit(), false, ent_user)) {
+        lu_ent_set_string(ent_user, LU_GECOS, name.toLocal8Bit());
+        if (lu_user_add(context, ent_user, &error)) {
+            if (addUserToGroups(user)) {
+                if (makeHome(user)) {
+                    emit userAdded(user, name);
+                } else {
+                    QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Creating user home failed"));
+                }
+            } else {
+                QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Adding user to groups failed"));
+            }
+        } else {
+            qWarning() << "Adding user failed:" << lu_strerror(error);
+            QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, lu_strerror(error)));
+        }
+    } else {
+        qWarning() << "Getting user defaults failed";
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Getting user defaults failed"));
+    }
+
+    lu_ent_free(ent_user);
+    lu_end(context);
+}
+
+void SailfishUserManager::removeUser(const QString& user, const QDBusMessage& message)
+{
+    m_timer->start();
+
+    if (!removeHome(user)) {
+        QDBusConnection::systemBus().send(message.createErrorReply(QDBusError::Failed, "Removing user home failed"));
+        return;
```

Comment:
If both add and remove user functions bailout on first sign of troubles and leave semi valid user data behind, can we / how likely it is that we could end up in situation where some user names are permanently unusable?

Reply:
We'll figure out rollbacks later if needed.